ruby/lib/google/protobuf/ffi/descriptor.rb

```diff
@@ -0,0 +1,765 @@
+                message_descriptor = field_descriptor.subtype
+                key_field_def = Google::Protobuf::FFI.get_field_by_number(message_descriptor, 1)
+                key_field_type = Google::Protobuf::FFI.get_type(key_field_def)
+
+                value_field_def = Google::Protobuf::FFI.get_field_by_number(message_descriptor, 2)
+                value_field_type = Google::Protobuf::FFI.get_type(value_field_def)
+
+                message_value = Google::Protobuf::FFI.get_message_value(msg, field_descriptor)
+                iter = ::FFI::MemoryPointer.new(:size_t, 1)
+                iter.write(:size_t, Google::Protobuf::FFI::Upb_Map_Begin)
+                key_value_pairs = []
+                while Google::Protobuf::FFI.map_next(message_value[:map_val], iter) do
+                  iter_size_t = iter.read(:size_t)
+                  key_message_value = Google::Protobuf::FFI.map_key(message_value[:map_val], iter_size_t)
+                  value_message_value = Google::Protobuf::FFI.map_value(message_value[:map_val], iter_size_t)
+                  key_string = convert_upb_to_ruby(key_message_value, key_field_type).inspect
+                  value_string = inspect_field(value_field_def, value_field_type, value_message_value)
+                  key_value_pairs << "#{key_string}=>#{value_string}"
+                end
+                field_output << "#{field_descriptor.name}: {#{key_value_pairs.join(", ")}}"
+              elsif field_descriptor.repeated?
+                # TODO(jatl) Adapted - from repeated_field#each - can this be refactored to reduce echo?
+                repeated_field_output = []
+                message_value = Google::Protobuf::FFI.get_message_value(msg, field_descriptor)
+                array = message_value[:array_val]
+                n = array.null? ? 0 : Google::Protobuf::FFI.array_size(array)
+                0.upto(n - 1) do |i|
+                  element = Google::Protobuf::FFI.get_msgval_at(array, i)
+                  repeated_field_output << inspect_field(field_descriptor, field_descriptor.send(:c_type), element)
+                end
+                field_output << "#{field_descriptor.name}: [#{repeated_field_output.join(", ")}]"
+              else
+                message_value = Google::Protobuf::FFI.get_message_value msg, field_descriptor
+                rendered_value = inspect_field(field_descriptor, field_descriptor.send(:c_type), message_value)
+                field_output << "#{field_descriptor.name}: #{rendered_value}"
+              end
+            end
+            "<#{name}: #{field_output.join(', ')}>"
+          end
+
+          ##
+          # Gets a field of this message identified by the argument definition.
+          #
+          # @param field [FieldDescriptor] Descriptor of the field to get
+          def get_field(field, unwrap: false)
```

Comment:
The code currently switches back and forth between "static" methods `def self.foo()` and regular methods `def foo()`. This makes the code somewhat hard to follow. Could we have it organized so it mostly puts all the methods of one kind together?

Reply:
Done